src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationAnalysisValueProvider.cs

```diff
@@ -42,7 +42,18 @@ internal bool TryGetValue(TKey key, out TValue value)
             // Store the value for the lifetime of the compilation.
             lock (_valueMap)
             {
-                _valueMap[key] = value;
+                // Check if another thread already stored the computed value.
+                TValue storedValue;
+                if (_valueMap.TryGetValue(key, out storedValue))
+                {
+                    // If so, we return the stored value.
+                    value = storedValue;
```

Comment:
I can add a defensive test where multiple threads try to fetch the value from the same provider for a specific compilation.